File: packages/snow_draw_core/lib/draw/types/element_style.dart
import 'package:meta/meta.dart';

import 'draw_color.dart';

enum StrokeStyle { solid, dashed, dotted }

enum FillStyle { solid, line, crossLine }

enum HighlightShape { rectangle, ellipse }

enum CanvasFilterType { mosaic, gaussianBlur, grayscale, inversion }

enum ArrowType { straight, curved, elbow }

enum ArrowheadStyle {
  none,
  standard,
  triangle,
  square,
  circle,
  diamond,
  invertedTriangle,
  verticalLine,
}

enum TextHorizontalAlign { left, center, right }

enum TextVerticalAlign { top, center, bottom }

@immutable
class ElementStyleUpdate {
  const ElementStyleUpdate({
    this.color,
    this.fillColor,
    this.strokeWidth,
    this.strokeStyle,
    this.fillStyle,
    this.highlightShape,
    this.filterType,
    this.filterStrength,
    this.cornerRadius,
    this.arrowType,
    this.startArrowhead,
    this.endArrowhead,
    this.fontSize,
    this.fontFamily,
    this.textAlign,
    this.verticalAlign,
    this.textStrokeColor,
    this.textStrokeWidth,
    this.serialNumber,
  });

  final DrawColor? color;
  final DrawColor? fillColor;
  final double? strokeWidth;
  final StrokeStyle? strokeStyle;
  final FillStyle? fillStyle;
  final HighlightShape? highlightShape;
  final CanvasFilterType? filterType;
  final double? filterStrength;
  final double? cornerRadius;
  final ArrowType? arrowType;
  final ArrowheadStyle? startArrowhead;
  final ArrowheadStyle? endArrowhead;
  final double? fontSize;
  final String? fontFamily;
  final TextHorizontalAlign? textAlign;
  final TextVerticalAlign? verticalAlign;
  final DrawColor? textStrokeColor;
  final double? textStrokeWidth;
  final int? serialNumber;

  bool get isEmpty =>
      color == null &&
      fillColor == null &&
      strokeWidth == null &&
      strokeStyle == null &&
      fillStyle == null &&
      highlightShape == null &&
      filterType == null &&
      filterStrength == null &&
      cornerRadius == null &&
      arrowType == null &&
      startArrowhead == null &&
      endArrowhead == null &&
      fontSize == null &&
      fontFamily == null &&
      textAlign == null &&
      verticalAlign == null &&
      textStrokeColor == null &&
      textStrokeWidth == null &&
      serialNumber == null;
}

File: packages/snow_draw_rust_engine/src/draw/types/element_style.rs
#![allow(dead_code)]

use crate::draw::types::draw_color::DrawColor;
use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum StrokeStyle {
    Solid,
    Dashed,
    Dotted,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum FillStyle {
    Solid,
    Line,
    CrossLine,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum HighlightShape {
    Rectangle,
    Ellipse,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum CanvasFilterType {
    Mosaic,
    GaussianBlur,
    Grayscale,
    Inversion,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum ArrowType {
    Straight,
    Curved,
    Elbow,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum ArrowheadStyle {
    None,
    Standard,
    Triangle,
    Square,
    Circle,
    Diamond,
    InvertedTriangle,
    VerticalLine,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum TextHorizontalAlign {
    Left,
    Center,
    Right,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum TextVerticalAlign {
    Top,
    Center,
    Bottom,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct ElementStyleUpdate {
    pub color: Option<DrawColor>,
    pub fill_color: Option<DrawColor>,
    pub stroke_width: Option<f64>,
    pub stroke_style: Option<StrokeStyle>,
    pub fill_style: Option<FillStyle>,
    pub highlight_shape: Option<HighlightShape>,
    pub filter_type: Option<CanvasFilterType>,
    pub filter_strength: Option<f64>,
    pub corner_radius: Option<f64>,
    pub arrow_type: Option<ArrowType>,
    pub start_arrowhead: Option<ArrowheadStyle>,
    pub end_arrowhead: Option<ArrowheadStyle>,
    pub font_size: Option<f64>,
    pub font_family: Option<String>,
    pub text_align: Option<TextHorizontalAlign>,
    pub vertical_align: Option<TextVerticalAlign>,
    pub text_stroke_color: Option<DrawColor>,
    pub text_stroke_width: Option<f64>,
    pub serial_number: Option<i64>,
}

impl ElementStyleUpdate {
    pub fn is_empty(&self) -> bool {
        self.color.is_none()
            && self.fill_color.is_none()
            && self.stroke_width.is_none()
            && self.stroke_style.is_none()
            && self.fill_style.is_none()
            && self.highlight_shape.is_none()
            && self.filter_type.is_none()
            && self.filter_strength.is_none()
            && self.corner_radius.is_none()
            && self.arrow_type.is_none()
            && self.start_arrowhead.is_none()
            && self.end_arrowhead.is_none()
            && self.font_size.is_none()
            && self.font_family.is_none()
            && self.text_align.is_none()
            && self.vertical_align.is_none()
            && self.text_stroke_color.is_none()
            && self.text_stroke_width.is_none()
            && self.serial_number.is_none()
    }
}
